packages/compiler/test/cli/watch/watch.test.ts

```diff
@@ -0,0 +1,121 @@
+import { deepStrictEqual } from "assert";
+import { WatchEventType, mkdirSync } from "fs";
+import { appendFile, mkdir, rm } from "fs/promises";
+import { dirname } from "path";
+import { fileURLToPath } from "url";
+import { ProjectWatcher, createWatcher } from "../../../src/core/cli/watch.js";
+import { getDirectoryPath, resolvePath } from "../../../src/index.js";
+
+const __dirname = dirname(fileURLToPath(import.meta.url));
+
+const fixtureRoot = resolvePath(__dirname, "../../../../temp/test/cli/watcher");
+
+function fixturePath(path: string) {
+  return resolvePath(fixtureRoot, path);
+}
+
+class FixtureFS {
+  async modify(path: string, delay?: number): Promise<string> {
+    const realPath = fixturePath(path);
+    await this.#maybeDelay(async () => {
+      mkdirSync(getDirectoryPath(realPath), { recursive: true });
+      await appendFile(realPath, new Date().toISOString());
+    }, delay);
+    return realPath;
+  }
+
+  async rm(path: string, delay?: number): Promise<string> {
+    const realPath = fixturePath(path);
+    await this.#maybeDelay(async () => {
+      mkdirSync(getDirectoryPath(realPath), { recursive: true });
+      await rm(realPath);
+    }, delay);
+    return realPath;
+  }
+
+  #delayTimers: NodeJS.Timeout[] = [];
+  #maybeDelay(fn: () => Promise<void>, delay: number | undefined) {
+    if (delay) {
+      return new Promise((resolve, reject) => {
+        this.#delayTimers.push(
+          setTimeout(() => {
+            fn().then(resolve).catch(reject);
+          }, delay)
+        );
+      });
+    } else {
+      return fn();
+    }
+  }
+
+  clearDelayTimers() {
+    this.#delayTimers.forEach(clearTimeout);
+    this.#delayTimers.length = 0;
+  }
+}
+
+describe("compiler: node host", () => {
+  before(async () => {
+    try {
+      await rm(fixtureRoot, { recursive: true });
+    } catch {}
+    await mkdir(fixtureRoot, { recursive: true });
+  });
+
+  let watcher: ProjectWatcher;
+  let fixtures: FixtureFS;
+  let changes: [WatchEventType, string][];
+
+  beforeEach(() => {
+    fixtures = new FixtureFS();
+    changes = [];
+    watcher = createWatcher((evt, name) => {
+      changes.push([evt, name]);
+    });
+  });
+
+  afterEach(() => {
+    watcher.close();
+    fixtures.clearDelayTimers();
+  });
+
+  it("should watch a single file and keep watching", async () => {
+    const file1 = await fixtures.modify("single/file1.txt");
+    watcher.updateWatchedFiles([file1]);
+    deepStrictEqual(changes, [], "Should not report change in initial load.");
+    await fixtures.modify("single/file1.txt", 100);
+
+    await delay(100);
```

Comment:
Depending on delays to indicate the end of a task feels like a potential source of test flakiness.  Is there another mechanism we can use to ensure the modify function has completed its work?

Reply:
yeah I am not sure what is the best here, saw some test from node-watch just basically waiting until it regirsterd all the modifications so instead it would timeout which I find confusing.